### ISSUE TITLE
Check if affiliation matching job needs fixing 

### ISSUE THREAD

**Issue by przemyslawjacewicz** [openaire/iis#1249]
According to affiliation matching documentation the job uses different versions of `AffOrgMatcher` to match affiliations to organizations. If a match is found by more then one matcher then the match with max match strength is chosen. Within matchers the job uses `AffOrgJoiner` to bucket affiliations and organizations and `AffOrgMatchVoter` to verify matching. If a match is verified by more then one voter then a custom algorithm is applied to reduce all match strengths to a single value. 

However actual job implementation is different. Choosing a match from matcher results is ok - matchers results are reduced and the best match is chosen as the resulting one:
```java
//eu.dnetlib.iis.wf.affmatching.AffMatchingService
        JavaRDD<AffMatchResult> allMatchedAffOrgs = allMatchedAffOrgsWithKey
                .reduceByKey((r1, r2) -> affMatchResultChooser.chooseBetter(r1, r2))
                .values();
```
 
 Reduced matcher results are then reduced the second time to produce results with match strengths recalculated using the algorithm:
 ```java
//eu.dnetlib.iis.wf.affmatching.write.IisAffMatchResultWriter 
         JavaRDD<MatchedOrganization> matchedOrganizations = matchedAffOrgs.map(affOrgMatch -> affMatchResultConverter.convert(affOrgMatch));
        
        
        JavaRDD<MatchedOrganization> distinctMatchedOrganizations = matchedOrganizations
                .keyBy(match -> new Tuple2<>(match.getDocumentId(), match.getOrganizationId()))
                .reduceByKey((match1, match2) -> matchedOrgStrengthRecalculation.recalculateStrength(match1, match2))
                .values();
 ```
 This reduction is a noop because it process already reduced affilation match results. Another thing is that the algorithm for recalculating match strength for matches found by more then one voter is different:
 ```java
 //eu.dnetlib.iis.wf.affmatching.match.AffOrgMatchStrengthRecalculator
     public AffMatchResult recalculateMatchStrength(AffMatchResult affMatchResult, AffOrgMatchVoter voter) { 

        Preconditions.checkNotNull(affMatchResult);
        Preconditions.checkNotNull(voter);
        
        float matchStrength = affMatchResult.getMatchStrength();
        
        if (voter.voteMatch(affMatchResult.getAffiliation(), affMatchResult.getOrganization())) {
            
            matchStrength = (matchStrength + voter.getMatchStrength()) - matchStrength * voter.getMatchStrength(); // bear in mind: the strengths are less that one
            
        } 
        
        return new AffMatchResult(affMatchResult.getAffiliation(), affMatchResult.getOrganization(), matchStrength);
        

    }
 ```
 According to documentation each confirmed match should decrease the gap to 1 (max match strength) by voter match strength. The above implementation seems different. 
 
 We should check/fix two things within this task
 * fix/remove/move the second aggregation of matchers results 
 * verify/fix/change the algorithm for recalculation of match strength for matches found by more then one voter

Any changes should be checked in terms of output data to see how output data change. Hopefully we should not change the output drastically.

**Comment by przemyslawjacewicz** [comment]
A more in-depth analysis showed that affiliation matching job is okay
* the second reduction is necessary to produce unique match between **document and organization** - the first reduction produces unique match between **affiliation (identified by document id and position) and organization**, so for example the first reduction will produce records like this
```json
{"affiliation":{"documentId":"DOC_3","position":2,"organizationName":"department demography, unversity california at berkeley","countryName":"","countryCode":""},"organization":{"id":"20|UCB","name":"university california at berkeley","shortName":"uc berkeley","countryName":"united states","countryCode":"us","websiteUrl":"","alternativeNames":[]},"matchStrength":1.0}
{"affiliation":{"documentId":"DOC_3","position":1,"organizationName":"department molecular and cell biology, university california at berkeley","countryName":"united states","countryCode":"us"},"organization":{"id":"20|UCB","name":"university california at berkeley","shortName":"uc berkeley","countryName":"united states","countryCode":"us","websiteUrl":"","alternativeNames":[]},"matchStrength":1.0}
{"affiliation":{"documentId":"DOC_3","position":3,"organizationName":"interdisciplinary centre mathematical and computer modelling","countryName":"","countryCode":""},"organization":{"id":"20|ICMUW","name":"interdisciplinary centre mathematical and computer modelling","shortName":"icm","countryName":"poland","countryCode":"pl","websiteUrl":"icm.edu.pl","alternativeNames":[]},"matchStrength":0.9999931}
```
note position in affiliation struct and the second reduction will produce records like this
```json
{"documentId":"DOC_3","organizationId":"20|UCB","matchStrength":1.0}
{"documentId":"DOC_3","organizationId":"20|ICMUW","matchStrength":0.9999931}
 ```
The algorithm for recalculating match strength from multiple matches is used for the second reduction to produce a single value for match strength.

* the implementation of recalculation algorithm that seemed wrong is in fact a proper implementation of the algorithm for recalculating match strength, although a bit convoluted

So in summary this issue can be closed.
